src/virtio-net.c

```diff
@@ -0,0 +1,281 @@
+
+    if (write(dev->irqfd, &n, sizeof(n)) < 0)
+        throw_err("Failed to write the irqfd");
+}
+
+void virtio_net_complete_request_rx(struct virtq *vq)
+{
+    struct virtio_net_dev *dev = (struct virtio_net_dev *) vq->dev;
+    vm_t *v = container_of(dev, vm_t, virtio_net_dev);
+    struct vring_packed_desc *desc;
+
+    while ((desc = virtq_get_avail(vq)) != NULL) {
+        uint8_t *data = vm_guest_to_host(v, desc->addr);
+        struct virtio_net_hdr_v1 *virtio_hdr =
+            (struct virtio_net_hdr_v1 *) data;
+        memset(virtio_hdr, 0, sizeof(struct virtio_net_hdr_v1));
+
+        virtio_hdr->num_buffers = 1;
+
+        size_t virtio_header_len = sizeof(struct virtio_net_hdr_v1);
+        ssize_t read_bytes = read(dev->tapfd, data + virtio_header_len,
+                                  desc->len - virtio_header_len);
+        if (read_bytes < 0) {
+            vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_DISABLE;
+            return;
+        }
+        desc->len = virtio_header_len + read_bytes;
+
+        desc->flags ^= (1ULL << VRING_PACKED_DESC_F_USED);
+        dev->virtio_pci_dev.config.isr_cap.isr_status |= VIRTIO_PCI_ISR_QUEUE;
+        return;
+    }
+    vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_DISABLE;
+    return;
+}
+
+void virtio_net_complete_request_tx(struct virtq *vq)
+{
+    struct virtio_net_dev *dev = (struct virtio_net_dev *) vq->dev;
+    vm_t *v = container_of(dev, vm_t, virtio_net_dev);
+    struct vring_packed_desc *desc;
+    while ((desc = virtq_get_avail(vq)) != NULL) {
+        uint8_t *data = vm_guest_to_host(v, desc->addr);
+        size_t virtio_header_len = sizeof(struct virtio_net_hdr_v1);
+
+        if (desc->len < virtio_header_len) {
+            vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_DISABLE;
+            return;
+        }
+
+        uint8_t *actual_data = data + virtio_header_len;
+        size_t actual_data_len = desc->len - virtio_header_len;
+
+        struct iovec iov[1];
+        iov[0].iov_base = actual_data;
+        iov[0].iov_len = actual_data_len;
+
+        ssize_t write_bytes = writev(dev->tapfd, iov, 1);
+        if (write_bytes < 0) {
+            vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_DISABLE;
+            return;
+        }
+        desc->flags ^= (1ULL << VRING_PACKED_DESC_F_USED);
+        dev->virtio_pci_dev.config.isr_cap.isr_status |= VIRTIO_PCI_ISR_QUEUE;
+        return;
+    }
+    vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_DISABLE;
+    return;
+}
+
+static struct virtq_ops virtio_net_ops[VIRTIO_NET_VIRTQ_NUM] = {
+    [VIRTQ_RX] =
+        {
+            .enable_vq = virtio_net_enable_vq_rx,
+            .complete_request = virtio_net_complete_request_rx,
+            .notify_used = virtio_net_notify_used_rx,
+        },
+    [VIRTQ_TX] = {.enable_vq = virtio_net_enable_vq_tx,
+                  .complete_request = virtio_net_complete_request_tx,
+                  .notify_used = virtio_net_notify_used_tx}};
```

Comment:
Replace `.notify_used = virtio_net_notify_used_tx}};` with `.notify_used = virtio_net_notify_used_tx},};` and re-run `clang-format`.